### ISSUE TITLE
Errors in Mpeg4VcmVideoEncoder Issues

### ISSUE THREAD

**Issue by Ken98045** [baSSiLL/SharpAvi#57]
Creating a AddMpeg4VcmVideoStream with Xvid - Divide by zero error.

I'm able to create a stream and get down to WriteFrame using x264vfw.  It errors out with very strange errors:

System.ArgumentOutOfRangeException
  HResult=0x80131502
  Message=A positive number is expected. (Parameter 'length')
  Source=SharpAvi
  StackTrace:
   at SharpAvi.Output.EncodingVideoStreamWrapper.WriteFrame(Boolean isKeyFrame, Byte[] frameData, Int32 startIndex, Int32 length)
   at SharpAvi.Output.AsyncVideoStreamWrapper.<>c__DisplayClass2_0.<WriteFrame>b__0()
   at SharpAvi.Utilities.SequentialInvoker.Invoke(Action action)
   at Program.<Main>$(String[] args) in ......

I look at the length input variable it is set to # bytes before the call.  After the call it is zero, and the passed in array is now null.  Neither of these should be possible since they are input only values.  It almost looks like an old style C stack or buffer flow issue.  Yes, I know that the bitmap bits are locked, but that still shouldn't be possible?

I have spent hours triple checking all the my input values and all your settings.  I have spent more hours looking over the available documentation.

**Comment by Ken98045** [comment]
I did compile your library and sample code.  The same problems are there too.  More Info follows:
x264vfw - SingleThreadEncoderWrapper line 103 - Encode frame returns 0 length for encoded data.
Xvid - Mpeg4VcmVidenEncoder.cs line 275 - some error in StartCompression - divide by 0

There may possibly be an issue with the versions of x264vfw and Xvid -- I downloaded them fresh today.  It is conceivable that there was something I missed in both, but....  It is possible the the Xvid install needs to be their $20 "pro" version.  I did not install that.

**Comment by baSSiLL** [comment]
Have you tried the "Zero Latency" setting of x264vfw? See issue #49 

Xvid is known to produce the DivideByZero when running in 64 bits. I've never managed to work this around. Should work in 32 bits though.

**Comment by Ken98045** [comment]
Thank!  That did it!

 

BTW, is this documented somewhere?  If not, could you.  This cost me over 6+ hours trying to figure out if I was doing something wrong.

**Comment by baSSiLL** [comment]
It's mentioned in the docs https://bassill.github.io/SharpAvi/using-video-encoders.html#mpeg-4
as well as in comments for the `Mpeg4VcmVideoEncoder` class.

**Comment by Ken98045** [comment]
Thanks again!  FYI: In my case the symptoms were somewhat different.  It wouldn’t encode even one frame.  I must have missed the tip though.  Maybe I’m just more blind than I thought.  I did start looking at the source code, but I didn’t look too deeply once I hit the exceptions.

 

I know you warned against Divx.  I saw that they had a new version so I thought I’d give it a try once I ran into a roadblock on vfw.
